tokio-pty-process/src/lib.rs

```diff
@@ -387,6 +387,122 @@ impl AsAsyncPtyFd for AsyncPtyMaster {
     }
 }
 
+/// Trait containing generalized methods for PTYs
+pub trait PtyMaster {
+    /// Return the full pathname of the slave device counterpart
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// extern crate tokio;
+    /// extern crate tokio_pty_process;
+    ///
+    /// use std::ffi::OsString;
+    /// use tokio::prelude::*;
+    /// use tokio_pty_process::{AsyncPtyMaster, PtyMaster};
+    ///
+    /// struct PtsName<T: PtyMaster>(T);
+    ///
+    /// impl<T: PtyMaster> Future for PtsName<T> {
+    ///     type Item = OsString;
+    ///     type Error = std::io::Error;
+    ///
+    ///     fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
+    ///         self.0.ptsname()
+    ///     }
+    /// }
+    ///
+    /// fn main() {
+    ///     let master = AsyncPtyMaster::open().expect("Could not open the PTY");
+    ///
+    ///     let ptsname = PtsName(master).wait().expect("Could not get the ptsname");
+    ///
+    ///     println!("PTS name: {}", ptsname.to_string_lossy());
+    /// }
+    /// ```
+    fn ptsname(&self) -> Poll<OsString, io::Error>;
+
+    /// Resize the PTY
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// extern crate tokio;
+    /// extern crate tokio_pty_process;
+    /// extern crate libc;
+    ///
+    /// use tokio_pty_process::{AsyncPtyMaster, PtyMaster};
+    /// use tokio::prelude::*;
+    /// use std::ffi::OsString;
+    /// use libc::c_ushort;
+    /// struct Resize<T: PtyMaster> {
+    ///     pty: T,
+    ///     rows: c_ushort,
+    ///     cols: c_ushort,
+    /// }
+    ///
+    /// impl<T: PtyMaster> Future for Resize<T> {
+    ///     type Item = ();
+    ///     type Error = std::io::Error;
+    ///
+    ///     fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
+    ///         self.pty.resize(self.rows, self.cols)
+    ///     }
+    /// }
+    ///
+    /// fn main() {
+    ///     let master = AsyncPtyMaster::open().expect("Could not open the PTY");
+    ///
+    ///     Resize {
+    ///         pty: master,
+    ///         cols: 80,
+    ///         rows: 50,
+    ///     }
+    ///     .wait()
+    ///     .expect("Could not resize the PTY");
+    /// }
+    /// ```
+    fn resize(&self, rows: c_ushort, cols: c_ushort) -> Poll<(), io::Error>;
+}
+
+impl<T: AsAsyncPtyFd> PtyMaster for T {
+    fn ptsname(&self) -> Poll<OsString, io::Error> {
+        let mut buf: [libc::c_char; 512] = [0; 512];
+        let fd = try_ready!(self.as_async_pty_fd());
+
+        #[cfg(not(any(target_os = "macos", target_os = "freebsd")))]
```

Comment:
Here, can you please use the style that I came up with in `tokio-pty-process` [as shown here](https://github.com/pkgw/stund/blob/master/tokio-pty-process/src/lib.rs#L136-L144)? The key being:

```rust
const APPLY_NONBLOCK_AFTER_OPEN: bool = cfg!(target_os = "freebsd");
```

Turning the check into a named constant makes it clearer what the key difference is, *and* reduces tricky `#[cfg()]` logic, and of course it's all the same after the optimizer runs.

Reply:
Hmm, this does not seem to build on macos anymore:
```
error[E0425]: cannot find function `ptsname_r` in module `libc`
   --> tokio-pty-process/src/lib.rs:478:31
    |
478 |             if unsafe { libc::ptsname_r(fd, buf.as_mut_ptr(), buf.len()) } != 0 {
    |                               ^^^^^^^^^ did you mean `ptsname`?
```

Reply:
Ah, that's too bad! Oh well.